// moment/src/lib/duration/get.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.get = get;
exports.weeks = weeks;
exports.years = exports.months = exports.days = exports.hours = exports.minutes = exports.seconds = exports.milliseconds = void 0;

var _aliases = require("../units/aliases");

var _absFloor = _interopRequireDefault(require("../utils/abs-floor"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function get(units) {
  units = (0, _aliases.normalizeUnits)(units);
  return this.isValid() ? this[units + 's']() : NaN;
}

function makeGetter(name) {
  return function () {
    return this.isValid() ? this._data[name] : NaN;
  };
}

var milliseconds = makeGetter('milliseconds');
exports.milliseconds = milliseconds;
var seconds = makeGetter('seconds');
exports.seconds = seconds;
var minutes = makeGetter('minutes');
exports.minutes = minutes;
var hours = makeGetter('hours');
exports.hours = hours;
var days = makeGetter('days');
exports.days = days;
var months = makeGetter('months');
exports.months = months;
var years = makeGetter('years');
exports.years = years;

function weeks() {
  return (0, _absFloor["default"])(this.days() / 7);
}
