// moment/src/lib/moment/moment.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createUnix = createUnix;
exports.createInZone = createInZone;
Object.defineProperty(exports, "createLocal", {
  enumerable: true,
  get: function get() {
    return _local.createLocal;
  }
});
Object.defineProperty(exports, "createUTC", {
  enumerable: true,
  get: function get() {
    return _utc.createUTC;
  }
});
Object.defineProperty(exports, "createInvalid", {
  enumerable: true,
  get: function get() {
    return _valid.createInvalid;
  }
});
Object.defineProperty(exports, "isMoment", {
  enumerable: true,
  get: function get() {
    return _constructor.isMoment;
  }
});
Object.defineProperty(exports, "min", {
  enumerable: true,
  get: function get() {
    return _minMax.min;
  }
});
Object.defineProperty(exports, "max", {
  enumerable: true,
  get: function get() {
    return _minMax.max;
  }
});
Object.defineProperty(exports, "now", {
  enumerable: true,
  get: function get() {
    return _now.now;
  }
});
Object.defineProperty(exports, "momentPrototype", {
  enumerable: true,
  get: function get() {
    return _prototype["default"];
  }
});

var _local = require("../create/local");

var _utc = require("../create/utc");

var _valid = require("../create/valid");

var _constructor = require("./constructor");

var _minMax = require("./min-max");

var _now = require("./now");

var _prototype = _interopRequireDefault(require("./prototype"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function createUnix(input) {
  return (0, _local.createLocal)(input * 1000);
}

function createInZone() {
  return _local.createLocal.apply(null, arguments).parseZone();
}
